### PR TITLE
Duration must be a timedelta instead of string

### PR DESCRIPTION
**Pull request by rakshith91**
Fixes https://github.com/Azure/azure-sdk-for-python/issues/19250

### CODE REVIEW

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 1]
```diff
@@ -64,5 +66,8 @@ def construct_iso8601(start=None, end=None, duration=None):
         end = Serializer.serialize_iso(end)
         iso_str = duration + '/' + end
     else:
-        iso_str = duration
+        iso_str =  duration
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:69]:
```suggestion
        iso_str = duration
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 2]
```diff
@@ -50,6 +50,8 @@ def order_results(request_order, responses):
     return ordered
 
 def construct_iso8601(start=None, end=None, duration=None):
+    if duration is not None:
+        duration = _duration_to_iso8601(duration)
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]:
```suggestion
    try:
        duration = _duration_to_iso8601(duration)
    except AttributeError:
        pass
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 3]
```diff
@@ -50,6 +50,10 @@ def order_results(request_order, responses):
     return ordered
 
 def construct_iso8601(start=None, end=None, duration=None):
+    try:
+        duration = _duration_to_iso8601(duration)
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]:
```suggestion
    try:
        duration = 'PT{}S'.format(duration.total_seconds())
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 4]
```diff
@@ -66,3 +70,6 @@ def construct_iso8601(start=None, end=None, duration=None):
     else:
         iso_str = duration
     return iso_str
+
+def _duration_to_iso8601(duration):
+    return 'PT{}S'.format(duration.total_seconds())
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:75]:
```suggestion
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 5]
```diff
@@ -50,6 +50,10 @@ def order_results(request_order, responses):
     return ordered
 
 def construct_iso8601(start=None, end=None, duration=None):
+    try:
+        duration = 'PT{}S'.format(duration.total_seconds())
```
Comment (xiangyan99) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]: Should we use isinstance?

If user uses a class that has total_seconds property, is it valid?

Comment (xiangyan99) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]: What if duration is None in line 69?

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py [hunk 6]
```diff
@@ -49,7 +50,7 @@ def __init__(self, credential, **kwargs):
         self._query_op = self._client.query
 
     def query(self, workspace_id, query, duration=None, **kwargs):
-        # type: (str, str, str, Any) -> LogsQueryResults
+        # type: (str, str, timedelta, Any) -> LogsQueryResults
```
Comment (xiangyan99) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:53]:
```suggestion
        # type: (str, str, Optional[timedelta], Any) -> LogsQueryResults
```

Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_log_query_client.py:53]: i'll update all these :)

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py (hunk 5 above)
Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]: changed to `if duration is not None`

- for the second question - it's okay to be None

Comment (xiangyan99) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]: Just to confirm, you mean

```py
end = Serializer.serialize_iso(end)
iso_str = None + '/' + end
```
is valid?

Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:54]: oops - line 69-71 got changed when i made the changes :)
nope = `iso_str = None + '/' + end` is not a valid format - although end must be provided with a duration
- it's worth improving the message :)

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 7]
```diff
@@ -61,6 +63,8 @@ def construct_iso8601(start=None, end=None, duration=None):
         else:
             raise ValueError("Start time must be provided aling with duration or end time.")
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:64]:
```suggestion
            raise ValueError("Start time must be provided along with duration or end time.")
```

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 8]
```diff
@@ -61,6 +63,8 @@ def construct_iso8601(start=None, end=None, duration=None):
...
     elif end is not None:
+        if not duration:
+            raise ValueError("End time must be provided aling with duration or start time.")
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:67]:
```suggestion
            raise ValueError("End time must be provided along with duration or start time.")
```

File: sdk/monitor/azure-monitor-query/tests/test_logs_timespans.py [hunk 9]
```diff
@@ -83,10 +85,33 @@ def test_query_duration_only():
...
+
+    assert construct_iso8601(duration=d1) == 'PT86400.0S'
+    assert construct_iso8601(duration=d2) == 'PT604800.0S'
+    assert construct_iso8601(duration=d3) == 'PT2160000.0S'
+    assert construct_iso8601(duration=d4) == 'PT10.0S'
+    assert construct_iso8601(duration=d5) == 'PT0.001S'
+    assert construct_iso8601(duration=d5) == 'PT0.001S'
+    assert construct_iso8601(duration=d7) == 'PT172800.0S'
+
+    with pytest.raises(ValueError, match="End time must be provided aling with duration or start time."):
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/test_logs_timespans.py:113]:
```suggestion
    with pytest.raises(ValueError, match="End time must be provided along with duration or start time."):
```

File: sdk/monitor/azure-monitor-query/tests/test_logs_timespans.py [hunk 10]
```diff
@@ -83,10 +85,33 @@ def test_query_duration_only():
...
+        construct_iso8601(end=datetime.now(UTC()))
+
+    with pytest.raises(ValueError, match="Start time must be provided aling with duration or end time."):
```
Comment (rakshith91) [inline sdk/monitor/azure-monitor-query/tests/test_logs_timespans.py:116]:
```suggestion
    with pytest.raises(ValueError, match="Start time must be provided along with duration or end time."):
```

File: sdk/monitor/azure-monitor-query/README.md [hunk 11]
```diff
@@ -245,7 +247,7 @@ response = client.query(
     metrics_uri,
     metric_names=["PublishSuccessCount"],
     start_time=datetime(2021, 5, 25),
-    duration='P1D'
+        duration=timedelta(days=1),
```
Comment (xiangyan99) [inline sdk/monitor/azure-monitor-query/README.md:250]: indent?

File: sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py [hunk 12]
```diff
@@ -59,8 +61,10 @@ def construct_iso8601(start=None, end=None, duration=None):
         elif duration is not None:
```
Comment (xiangyan99) [inline sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:61]: What's the behavior if user provides start, end and duration?

Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:61]: we would take the start and end time - we have documented that all three should not be provided - on retrospect, let me warn here too

Comment (xiangyan99) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:61]: In fact my question is if three are provided, we would silently ignore duration (current behavior) or give error? :)

Comment (rakshith91) [reply sdk/monitor/azure-monitor-query/azure/monitor/query/_helpers.py:61]: we now give error :)

**Review by xiangyan99 — APPROVED** [review]

**Pull request merged**